---
title: "gene_regions"
format: html
---

```{r}
#| message: false
library(BiocIO)
library(readr)
library(rtracklayer)
```

```{r}
gtf_file <- "/ceph/project/goodwin/albrecht/cellranger_mkref/results/cellranger_mkref/genes/genes.gtf.gz"
gtf_gene_data <- import(gtf_file, feature.type = "gene")
length(gtf_gene_data)
```

```{r}
table(seqnames(gtf_gene_data))
```

```{r}
exclude_chr <- seqnames(gtf_gene_data) %in% c("X", "Y", "mitochondrion_genome")
gtf_gene_filtered <- gtf_gene_data[!exclude_chr]
length(gtf_gene_filtered)
```

```{r}
table(gtf_gene_filtered$gene_biotype)
```

```{r}
exclude_chr <- gtf_gene_filtered$gene_biotype %in% c("ncRNA", "pseudogene", "rRNA", "snoRNA", "snRNA", "transposable_element", "tRNA")
gtf_gene_filtered <- gtf_gene_data[!exclude_chr]
length(gtf_gene_filtered)
```


```{r}
gene_regions_bed <- tibble(
  chr = as.character(seqnames(gtf_gene_filtered)),
  start = start(gtf_gene_filtered),
  end = end(gtf_gene_filtered)
)
gene_regions_bed
```

```{r}
write_tsv(x = gene_regions_bed, file = "gene_regions.bed", col_names = FALSE)
```

